### PR TITLE
#28 Retrieve reservations - initial test and stub

### PR DESCRIPTION
**Pull request by carlosmiranda**
#28: Added API endpoint method stub for retrieving reservations in JSON format. Added test, and puzzle for continuing implemenation.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#254](https://github.com/timelesslounge/timelessis/issues/254) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#254](https://github.com/timelesslounge/timelessis/issues/254) is assigned to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/eb636d39-da16-431b-b8a8-6145aa1aad35); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: tests/test_reservation.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from datetime import datetime
+from http import HTTPStatus
 from timeless.restaurants.models import ReservationStatus, Reservation
```
Comment (vladarefiev) [inline tests/test_reservation.py]: @carlosmiranda according to pep8 here should be blank line between 2 and 3 lines

File: timeless/reservations/views.py [hunk 2]
```diff
@@ -78,6 +80,25 @@ def list_reservations(reservations):
         "restaurants/tables/list.html", reservations=reservations
     )
 
+@bp.route("/retrieve", methods=("GET", "POST"))
+def retrieve_reservations():
```
Comment (vladarefiev) [inline timeless/reservations/views.py]: @carlosmiranda according to task so far I can see that endpoint should be linked to a specific company, then we should provide `company_id` to view.

File: timeless/reservations/views.py [hunk 3]
```diff
@@ -78,6 +80,25 @@ def list_reservations(reservations):
         "restaurants/tables/list.html", reservations=reservations
     )
 
+@bp.route("/retrieve", methods=("GET", "POST"))
```
Comment (vladarefiev) [inline timeless/reservations/views.py]: @carlosmiranda why do we need both `get` and `post`?

Comment (vladarefiev) [inline timeless/reservations/views.py]: @carlosmiranda @emilianodellacasa Here may be a confusion. We have here mix of a lot end-points. We already use `/api` prefix for API's
here is an example - https://github.com/timelesslounge/timelessis/blob/master/timeless/__init__.py#L94

File: tests/test_reservation.py [hunk 4]
```diff
@@ -44,4 +45,9 @@ def test_reservation_duration():
         comment=comment,
         status=status
     )
-    assert new_reservation.duration() == end_time - start_time
\ No newline at end of file
+    assert new_reservation.duration() == end_time - start_time
+
+def test_retrieve(client):
+    response = client.post("/reservations/retrieve", data={"location" : 1, "date" : datetime(1, 1, 1)})
+    assert (response.status_code == HTTPStatus.OK
```
Comment (vladarefiev) [inline tests/test_reservation.py]: @carlosmiranda one assert for one test pls, we've discussed it in telegram channel.
`assert response.status_code == HTTPStatus.OK`
`assert response.is_json`

**Comment by vladarefiev** [comment]
@carlosmiranda hey, good, just a few comments

**Comment by carlosmiranda** [comment]
@vladarefiev see newest commit. Sorry for the delay, I had some issues with my Ubuntu machine

File: tests/test_reservation.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from datetime import datetime
+from http import HTTPStatus
+
+
+def test_retrieve_status(client):
+    response = client.post("/api/reservations/", data={"location" : 1, "date" : datetime(1, 1, 1)})
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_retrieve_returns_json(client):
```
Comment (vladarefiev) [inline tests/test_reservation.py]: @carlosmiranda I believe you didn't understand my previous comment clearly :)
Let's do just one test in this way

```
def test_retrieve_reservation_list(client):
    response = client.post("/api/reservations/", data={"location" : 1, "date" : datetime(1, 1, 1)})
    assert response.status_code == HTTPStatus.OK
    assert response.is_json
```

File: timeless/reservations/views.py [hunk 6]
```diff
@@ -44,6 +46,31 @@ class CommentView(CrudAPIView):
     url_lookup = "comment_id"
 
 
+class JsonView(views.MethodView):
```
Comment (vladarefiev) [inline timeless/reservations/views.py]: @carlosmiranda I don't particularly like this name, it sounds like generic one.
Let's call it ReservationListView or smth like that

File: timeless/reservations/views.py [hunk 7]
```diff
@@ -44,6 +46,31 @@ class CommentView(CrudAPIView):
...
+    """ Reservation JSON API """
+
+    def post(self):
```
Comment (vladarefiev) [inline timeless/reservations/views.py]: @carlosmiranda I think it should be `get`

File: timeless/reservations/views.py [hunk 8]
```diff
@@ -44,6 +46,31 @@ class CommentView(CrudAPIView):
...
+            "1": {
+                "start_time" : datetime(1, 1, 1),
+                "end_time" : datetime(1, 1, 1),
+                "customer_id" : 1,
+                "num_of_persons" : 1,
+                "comment" : "Test",
+                "status" : 2
+            }
+        }
+        return jsonify(reservations_json)
```
Comment (vladarefiev) [inline timeless/reservations/views.py]: @carlosmiranda @emilianodellacasa So far as I can see here should be a list of objects.
```
{'items': [{}, {}, ...]}
```

**Comment by vladarefiev** [comment]
@carlosmiranda take a look pls, and check `codeclimate`

File: timeless/reservations/views.py [hunk 9]
```diff
@@ -44,6 +46,31 @@ class CommentView(CrudAPIView):
...
+    def post(self):
+        """Retrieve reservations based on login, location, and date.
+            @todo #28:30min Implement actual fetching of reservations. We need to
+             return a JSON list, filtered by reservations based on location and
+             date. We also need to password protect this API, and filter only
+             those belonging to the specific company,
+        :param company_id:
+        :return:
+        """
+        reservations_json = {
```
Comment (vladarefiev) [inline timeless/reservations/views.py]: @carlosmiranda @emilianodellacasa 
I think we should return puzzle for implementation serialization and deserialization objects using this lib - https://flask-marshmallow.readthedocs.io/en/latest/ if we provide API.
if @emilianodellacasa agree with it, @carlosmiranda pls leave a puzzle here

Comment (emilianodellacasa) [reply timeless/reservations/views.py]: @vladarefiev I think we should

**Comment by carlosmiranda** [comment]
@vladarefiev see latest changes. @emilianodellacasa how can I see Code Climate result, I am getting 403 error.

**Comment by emilianodellacasa** [comment]
@carlosmiranda now you should be able to see it

**Comment by vladarefiev** [comment]
@carlosmiranda Thanks, now it's good for me. Fix codeclimate and let's merge

**Comment by carlosmiranda** [comment]
@emilianodellacasa still can't see code climate :(

**Comment by emilianodellacasa** [comment]
@carlosmiranda I have added you to the team and I can correctly see the code smells, have you tried cleaning your browser's cache?

**Comment by 0crat** [comment]
@vladarefiev[/z](https://www.0crat.com/u/vladarefiev) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job [#254](https://github.com/timelesslounge/timelessis/issues/254) is not in the agenda of @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), can't inspect

**Comment by 0crat** [comment]
The job [#254](https://github.com/timelesslounge/timelessis/issues/254) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by carlosmiranda** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/timelesslounge/timelessis/issues/254#issuecomment-465479894))

@carlosmiranda This is what I know about this job in [CF9LG1YCA](https://www.0crat.com/p/CF9LG1YCA), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#254](https://github.com/timelesslounge/timelessis/issues/254) is not in scope
  * The job is waiting quality review verdict by @ypshenychka[/z](https://www.0crat.com/u/ypshenychka) for 8min
  * @vladarefiev[/z](https://www.0crat.com/u/vladarefiev) will receive 15 minutes after quality review
  * Job [footprint](https://www.0crat.com/footprint/CF9LG1YCA?q=%7Bjob%3A%27gh%3Atimelesslounge%2Ftimelessis%23254%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev)

**Pull request merged**